### PR TITLE
settings: Support the fontconfig timestamp

### PR DESCRIPTION
**Pull request by matthiasclasen**
This is not a gsetting, so we treat it separately.
To make things easy for clients, we put it in a
fake org.gnome.fontconfig namespace.

The fontconfig monitor code here is copied from
gnome-settings-daemon.

### CODE REVIEW

File: src/settings.c [hunk 1]
```diff
@@ -138,31 +153,32 @@ settings_handle_read (XdpSettings           *object,
...
-      else
-        {
-          g_debug ("Attempted to read unknown key %s", arg_key);
-          g_dbus_method_invocation_return_error_literal (invocation, XDG_DESKTOP_PORTAL_ERROR,
-                                                         XDG_DESKTOP_PORTAL_ERROR_NOT_FOUND,
-                                                         _("Requested setting not found"));
+          g_dbus_method_invocation_return_value (invocation, g_variant_new ("(v)", variant));
+          return TRUE;
         }
     }
```
Comment (matthiasclasen) [inline src/settings.c]: I've turned around the logic here to use early returns for successful cases, and have a catch-all error case at the end.

File: src/settings.c [hunk 2]
```diff
@@ -238,6 +254,22 @@ init_settings_table (Settings   *self,
...
+fontconfig_changed (FcMonitor *monitor,
+                    Settings *self)
+{
+  const char *namespace = "org.gnome.fontconfig";
+  const char *key = "timestamp";
+  
+  g_debug ("Emitting changed for %s %s", namespace, key);
+
+  self->fontconfig_timestamp = time (NULL);
+
```
Comment (matthiasclasen) [inline src/settings.c]: This is the same that gsd does: use use time()

File: src/settings.c [hunk 3]
```diff
@@ -138,31 +153,32 @@ settings_handle_read (XdpSettings           *object,
...
-      g_debug ("Attempted to read from unknown namespace %s", arg_namespace);
-      g_dbus_method_invocation_return_error_literal (invocation, XDG_DESKTOP_PORTAL_ERROR,
-                                                     XDG_DESKTOP_PORTAL_ERROR_NOT_FOUND,
-                                                     _("Requested setting not found"));
+      if (strcmp (arg_key, "timestamp") == 0)
+        {
+          g_dbus_method_invocation_return_value (invocation,
+                                                 g_variant_new ("(v)", g_variant_new_int32 (self->fontconfig_timestamp)));
+          return TRUE;
+        }
```
Comment (TingPing) [inline src/settings.c]: else error?

File: src/settings.c (hunk 2 above)
Comment (TingPing) [reply src/settings.c]: I'm admittedly ignorant where this timestamp is used, but would it be a breaking change to make it monotonic?

File: src/fc-monitor.c [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+                g_debug ("Got %-38s for %s: restarting fontconfig update", event_name, path);
+                self->state = UPDATE_RESTART;
+                break;
+
+        case UPDATE_RESTART:
+                g_debug ("Got %-38s for %s: waiting on fontconfig update", event_name, path);
+                break;
+        }
+
+	g_free (path);
```
Comment (TingPing) [inline src/fc-monitor.c]:
```suggestion
        g_free (path);
```

File: src/fc-monitor.c [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+}
+
+static void
+update_done (GObject *source_object G_GNUC_UNUSED,
+             GAsyncResult *result,
+             gpointer data)
+{
+        FcMonitor *self = FC_MONITOR (data);
+        gboolean restart = self->state == UPDATE_RESTART;
+        GError *error = NULL;
```
Comment (TingPing) [inline src/fc-monitor.c]: We already use `g_autoptr()` in this project so might as well use it throughout here.

File: src/settings.c [hunk 6]
```diff
@@ -36,6 +38,8 @@ struct _Settings
   XdpSettingsSkeleton parent_instance;
 
   GHashTable *settings;
+  FcMonitor *fontconfig_monitor;
```
Comment (TingPing) [inline src/settings.c]: Maybe not ideal to have the exact same `Fc` prefix as real fontconfig types.

File: src/settings.c [hunk 7]
```diff
@@ -121,6 +125,17 @@ settings_handle_read_all (XdpSettings           *object,
 
       g_variant_builder_add (&builder, "{s@a{sv}}", key, g_variant_dict_end (&dict));
     }
+
+  if (namespace_matches ("org.gnome.fontconfig", arg_namespace))
+    {
+      GVariantDict dict;
+
+      g_variant_dict_init (&dict, NULL);
+      g_variant_dict_insert_value (&dict, "timestamp", g_variant_new_int32 (self->fontconfig_timestamp));
```
Comment (TingPing) [inline src/settings.c]: Probably best to keep timestamps in int64.

Comment (matthiasclasen) [reply src/settings.c]: can't do that, since xsettings only knows string, bool and int, and we're still kinda limited by those constraints on the gtk side

Comment (TingPing) [reply src/settings.c]: The GTK MR for it already used int64: https://gitlab.gnome.org/GNOME/gtk/merge_requests/241

Even if it gets truncated by GTK in the short term might as well make the DBus API use the right type.

Comment (matthiasclasen) [reply src/settings.c]: I'd rather rename it. The thing is not a real timestamp in any way

Comment (matthiasclasen) [reply src/settings.c]: I've done that now

**Comment by hadess** [comment]
Any chance you could feed the changes to that copy/paste code back to g-s-d?

**Pull request merged**